<!DOCTYPE html>
<html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8"/>

<title>Description</title>

<script type="text/javascript">
window.onload = function() {
  var imgs = document.getElementsByTagName('img'), i, img;
  for (i = 0; i < imgs.length; i++) {
    img = imgs[i];
    // center an image if it is the only element of its parent
    if (img.parentElement.childElementCount === 1)
      img.parentElement.style.textAlign = 'center';
  }
};
</script>

<!-- Styles for R syntax highlighter -->
<style type="text/css">
   pre .operator,
   pre .paren {
     color: rgb(104, 118, 135)
   }

   pre .literal {
     color: #990073
   }

   pre .number {
     color: #099;
   }

   pre .comment {
     color: #998;
     font-style: italic
   }

   pre .keyword {
     color: #900;
     font-weight: bold
   }

   pre .identifier {
     color: rgb(0, 0, 0);
   }

   pre .string {
     color: #d14;
   }
</style>

<!-- R syntax highlighter -->
<script type="text/javascript">
var hljs=new function(){function m(p){return p.replace(/&/gm,"&amp;").replace(/</gm,"&lt;")}function f(r,q,p){return RegExp(q,"m"+(r.cI?"i":"")+(p?"g":""))}function b(r){for(var p=0;p<r.childNodes.length;p++){var q=r.childNodes[p];if(q.nodeName=="CODE"){return q}if(!(q.nodeType==3&&q.nodeValue.match(/\s+/))){break}}}function h(t,s){var p="";for(var r=0;r<t.childNodes.length;r++){if(t.childNodes[r].nodeType==3){var q=t.childNodes[r].nodeValue;if(s){q=q.replace(/\n/g,"")}p+=q}else{if(t.childNodes[r].nodeName=="BR"){p+="\n"}else{p+=h(t.childNodes[r])}}}if(/MSIE [678]/.test(navigator.userAgent)){p=p.replace(/\r/g,"\n")}return p}function a(s){var r=s.className.split(/\s+/);r=r.concat(s.parentNode.className.split(/\s+/));for(var q=0;q<r.length;q++){var p=r[q].replace(/^language-/,"");if(e[p]){return p}}}function c(q){var p=[];(function(s,t){for(var r=0;r<s.childNodes.length;r++){if(s.childNodes[r].nodeType==3){t+=s.childNodes[r].nodeValue.length}else{if(s.childNodes[r].nodeName=="BR"){t+=1}else{if(s.childNodes[r].nodeType==1){p.push({event:"start",offset:t,node:s.childNodes[r]});t=arguments.callee(s.childNodes[r],t);p.push({event:"stop",offset:t,node:s.childNodes[r]})}}}}return t})(q,0);return p}function k(y,w,x){var q=0;var z="";var s=[];function u(){if(y.length&&w.length){if(y[0].offset!=w[0].offset){return(y[0].offset<w[0].offset)?y:w}else{return w[0].event=="start"?y:w}}else{return y.length?y:w}}function t(D){var A="<"+D.nodeName.toLowerCase();for(var B=0;B<D.attributes.length;B++){var C=D.attributes[B];A+=" "+C.nodeName.toLowerCase();if(C.value!==undefined&&C.value!==false&&C.value!==null){A+='="'+m(C.value)+'"'}}return A+">"}while(y.length||w.length){var v=u().splice(0,1)[0];z+=m(x.substr(q,v.offset-q));q=v.offset;if(v.event=="start"){z+=t(v.node);s.push(v.node)}else{if(v.event=="stop"){var p,r=s.length;do{r--;p=s[r];z+=("</"+p.nodeName.toLowerCase()+">")}while(p!=v.node);s.splice(r,1);while(r<s.length){z+=t(s[r]);r++}}}}return z+m(x.substr(q))}function j(){function q(x,y,v){if(x.compiled){return}var u;var s=[];if(x.k){x.lR=f(y,x.l||hljs.IR,true);for(var w in x.k){if(!x.k.hasOwnProperty(w)){continue}if(x.k[w] instanceof Object){u=x.k[w]}else{u=x.k;w="keyword"}for(var r in u){if(!u.hasOwnProperty(r)){continue}x.k[r]=[w,u[r]];s.push(r)}}}if(!v){if(x.bWK){x.b="\\b("+s.join("|")+")\\s"}x.bR=f(y,x.b?x.b:"\\B|\\b");if(!x.e&&!x.eW){x.e="\\B|\\b"}if(x.e){x.eR=f(y,x.e)}}if(x.i){x.iR=f(y,x.i)}if(x.r===undefined){x.r=1}if(!x.c){x.c=[]}x.compiled=true;for(var t=0;t<x.c.length;t++){if(x.c[t]=="self"){x.c[t]=x}q(x.c[t],y,false)}if(x.starts){q(x.starts,y,false)}}for(var p in e){if(!e.hasOwnProperty(p)){continue}q(e[p].dM,e[p],true)}}function d(B,C){if(!j.called){j();j.called=true}function q(r,M){for(var L=0;L<M.c.length;L++){if((M.c[L].bR.exec(r)||[null])[0]==r){return M.c[L]}}}function v(L,r){if(D[L].e&&D[L].eR.test(r)){return 1}if(D[L].eW){var M=v(L-1,r);return M?M+1:0}return 0}function w(r,L){return L.i&&L.iR.test(r)}function K(N,O){var M=[];for(var L=0;L<N.c.length;L++){M.push(N.c[L].b)}var r=D.length-1;do{if(D[r].e){M.push(D[r].e)}r--}while(D[r+1].eW);if(N.i){M.push(N.i)}return f(O,M.join("|"),true)}function p(M,L){var N=D[D.length-1];if(!N.t){N.t=K(N,E)}N.t.lastIndex=L;var r=N.t.exec(M);return r?[M.substr(L,r.index-L),r[0],false]:[M.substr(L),"",true]}function z(N,r){var L=E.cI?r[0].toLowerCase():r[0];var M=N.k[L];if(M&&M instanceof Array){return M}return false}function F(L,P){L=m(L);if(!P.k){return L}var r="";var O=0;P.lR.lastIndex=0;var M=P.lR.exec(L);while(M){r+=L.substr(O,M.index-O);var N=z(P,M);if(N){x+=N[1];r+='<span class="'+N[0]+'">'+M[0]+"</span>"}else{r+=M[0]}O=P.lR.lastIndex;M=P.lR.exec(L)}return r+L.substr(O,L.length-O)}function J(L,M){if(M.sL&&e[M.sL]){var r=d(M.sL,L);x+=r.keyword_count;return r.value}else{return F(L,M)}}function I(M,r){var L=M.cN?'<span class="'+M.cN+'">':"";if(M.rB){y+=L;M.buffer=""}else{if(M.eB){y+=m(r)+L;M.buffer=""}else{y+=L;M.buffer=r}}D.push(M);A+=M.r}function G(N,M,Q){var R=D[D.length-1];if(Q){y+=J(R.buffer+N,R);return false}var P=q(M,R);if(P){y+=J(R.buffer+N,R);I(P,M);return P.rB}var L=v(D.length-1,M);if(L){var O=R.cN?"</span>":"";if(R.rE){y+=J(R.buffer+N,R)+O}else{if(R.eE){y+=J(R.buffer+N,R)+O+m(M)}else{y+=J(R.buffer+N+M,R)+O}}while(L>1){O=D[D.length-2].cN?"</span>":"";y+=O;L--;D.length--}var r=D[D.length-1];D.length--;D[D.length-1].buffer="";if(r.starts){I(r.starts,"")}return R.rE}if(w(M,R)){throw"Illegal"}}var E=e[B];var D=[E.dM];var A=0;var x=0;var y="";try{var s,u=0;E.dM.buffer="";do{s=p(C,u);var t=G(s[0],s[1],s[2]);u+=s[0].length;if(!t){u+=s[1].length}}while(!s[2]);if(D.length>1){throw"Illegal"}return{r:A,keyword_count:x,value:y}}catch(H){if(H=="Illegal"){return{r:0,keyword_count:0,value:m(C)}}else{throw H}}}function g(t){var p={keyword_count:0,r:0,value:m(t)};var r=p;for(var q in e){if(!e.hasOwnProperty(q)){continue}var s=d(q,t);s.language=q;if(s.keyword_count+s.r>r.keyword_count+r.r){r=s}if(s.keyword_count+s.r>p.keyword_count+p.r){r=p;p=s}}if(r.language){p.second_best=r}return p}function i(r,q,p){if(q){r=r.replace(/^((<[^>]+>|\t)+)/gm,function(t,w,v,u){return w.replace(/\t/g,q)})}if(p){r=r.replace(/\n/g,"<br>")}return r}function n(t,w,r){var x=h(t,r);var v=a(t);var y,s;if(v){y=d(v,x)}else{return}var q=c(t);if(q.length){s=document.createElement("pre");s.innerHTML=y.value;y.value=k(q,c(s),x)}y.value=i(y.value,w,r);var u=t.className;if(!u.match("(\\s|^)(language-)?"+v+"(\\s|$)")){u=u?(u+" "+v):v}if(/MSIE [678]/.test(navigator.userAgent)&&t.tagName=="CODE"&&t.parentNode.tagName=="PRE"){s=t.parentNode;var p=document.createElement("div");p.innerHTML="<pre><code>"+y.value+"</code></pre>";t=p.firstChild.firstChild;p.firstChild.cN=s.cN;s.parentNode.replaceChild(p.firstChild,s)}else{t.innerHTML=y.value}t.className=u;t.result={language:v,kw:y.keyword_count,re:y.r};if(y.second_best){t.second_best={language:y.second_best.language,kw:y.second_best.keyword_count,re:y.second_best.r}}}function o(){if(o.called){return}o.called=true;var r=document.getElementsByTagName("pre");for(var p=0;p<r.length;p++){var q=b(r[p]);if(q){n(q,hljs.tabReplace)}}}function l(){if(window.addEventListener){window.addEventListener("DOMContentLoaded",o,false);window.addEventListener("load",o,false)}else{if(window.attachEvent){window.attachEvent("onload",o)}else{window.onload=o}}}var e={};this.LANGUAGES=e;this.highlight=d;this.highlightAuto=g;this.fixMarkup=i;this.highlightBlock=n;this.initHighlighting=o;this.initHighlightingOnLoad=l;this.IR="[a-zA-Z][a-zA-Z0-9_]*";this.UIR="[a-zA-Z_][a-zA-Z0-9_]*";this.NR="\\b\\d+(\\.\\d+)?";this.CNR="\\b(0[xX][a-fA-F0-9]+|(\\d+(\\.\\d*)?|\\.\\d+)([eE][-+]?\\d+)?)";this.BNR="\\b(0b[01]+)";this.RSR="!|!=|!==|%|%=|&|&&|&=|\\*|\\*=|\\+|\\+=|,|\\.|-|-=|/|/=|:|;|<|<<|<<=|<=|=|==|===|>|>=|>>|>>=|>>>|>>>=|\\?|\\[|\\{|\\(|\\^|\\^=|\\||\\|=|\\|\\||~";this.ER="(?![\\s\\S])";this.BE={b:"\\\\.",r:0};this.ASM={cN:"string",b:"'",e:"'",i:"\\n",c:[this.BE],r:0};this.QSM={cN:"string",b:'"',e:'"',i:"\\n",c:[this.BE],r:0};this.CLCM={cN:"comment",b:"//",e:"$"};this.CBLCLM={cN:"comment",b:"/\\*",e:"\\*/"};this.HCM={cN:"comment",b:"#",e:"$"};this.NM={cN:"number",b:this.NR,r:0};this.CNM={cN:"number",b:this.CNR,r:0};this.BNM={cN:"number",b:this.BNR,r:0};this.inherit=function(r,s){var p={};for(var q in r){p[q]=r[q]}if(s){for(var q in s){p[q]=s[q]}}return p}}();hljs.LANGUAGES.cpp=function(){var a={keyword:{"false":1,"int":1,"float":1,"while":1,"private":1,"char":1,"catch":1,"export":1,virtual:1,operator:2,sizeof:2,dynamic_cast:2,typedef:2,const_cast:2,"const":1,struct:1,"for":1,static_cast:2,union:1,namespace:1,unsigned:1,"long":1,"throw":1,"volatile":2,"static":1,"protected":1,bool:1,template:1,mutable:1,"if":1,"public":1,friend:2,"do":1,"return":1,"goto":1,auto:1,"void":2,"enum":1,"else":1,"break":1,"new":1,extern:1,using:1,"true":1,"class":1,asm:1,"case":1,typeid:1,"short":1,reinterpret_cast:2,"default":1,"double":1,register:1,explicit:1,signed:1,typename:1,"try":1,"this":1,"switch":1,"continue":1,wchar_t:1,inline:1,"delete":1,alignof:1,char16_t:1,char32_t:1,constexpr:1,decltype:1,noexcept:1,nullptr:1,static_assert:1,thread_local:1,restrict:1,_Bool:1,complex:1},built_in:{std:1,string:1,cin:1,cout:1,cerr:1,clog:1,stringstream:1,istringstream:1,ostringstream:1,auto_ptr:1,deque:1,list:1,queue:1,stack:1,vector:1,map:1,set:1,bitset:1,multiset:1,multimap:1,unordered_set:1,unordered_map:1,unordered_multiset:1,unordered_multimap:1,array:1,shared_ptr:1}};return{dM:{k:a,i:"</",c:[hljs.CLCM,hljs.CBLCLM,hljs.QSM,{cN:"string",b:"'\\\\?.",e:"'",i:"."},{cN:"number",b:"\\b(\\d+(\\.\\d*)?|\\.\\d+)(u|U|l|L|ul|UL|f|F)"},hljs.CNM,{cN:"preprocessor",b:"#",e:"$"},{cN:"stl_container",b:"\\b(deque|list|queue|stack|vector|map|set|bitset|multiset|multimap|unordered_map|unordered_set|unordered_multiset|unordered_multimap|array)\\s*<",e:">",k:a,r:10,c:["self"]}]}}}();hljs.LANGUAGES.r={dM:{c:[hljs.HCM,{cN:"number",b:"\\b0[xX][0-9a-fA-F]+[Li]?\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\b\\d+(?:[eE][+\\-]?\\d*)?L\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\b\\d+\\.(?!\\d)(?:i\\b)?",e:hljs.IMMEDIATE_RE,r:1},{cN:"number",b:"\\b\\d+(?:\\.\\d*)?(?:[eE][+\\-]?\\d*)?i?\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\.\\d+(?:[eE][+\\-]?\\d*)?i?\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"keyword",b:"(?:tryCatch|library|setGeneric|setGroupGeneric)\\b",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\.\\.\\.",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\.\\.\\d+(?![\\w.])",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\b(?:function)",e:hljs.IMMEDIATE_RE,r:2},{cN:"keyword",b:"(?:if|in|break|next|repeat|else|for|return|switch|while|try|stop|warning|require|attach|detach|source|setMethod|setClass)\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"literal",b:"(?:NA|NA_integer_|NA_real_|NA_character_|NA_complex_)\\b",e:hljs.IMMEDIATE_RE,r:10},{cN:"literal",b:"(?:NULL|TRUE|FALSE|T|F|Inf|NaN)\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"identifier",b:"[a-zA-Z.][a-zA-Z0-9._]*\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"operator",b:"<\\-(?!\\s*\\d)",e:hljs.IMMEDIATE_RE,r:2},{cN:"operator",b:"\\->|<\\-",e:hljs.IMMEDIATE_RE,r:1},{cN:"operator",b:"%%|~",e:hljs.IMMEDIATE_RE},{cN:"operator",b:">=|<=|==|!=|\\|\\||&&|=|\\+|\\-|\\*|/|\\^|>|<|!|&|\\||\\$|:",e:hljs.IMMEDIATE_RE,r:0},{cN:"operator",b:"%",e:"%",i:"\\n",r:1},{cN:"identifier",b:"`",e:"`",r:0},{cN:"string",b:'"',e:'"',c:[hljs.BE],r:0},{cN:"string",b:"'",e:"'",c:[hljs.BE],r:0},{cN:"paren",b:"[[({\\])}]",e:hljs.IMMEDIATE_RE,r:0}]}};
hljs.initHighlightingOnLoad();
</script>



<style type="text/css">
body, td {
   font-family: sans-serif;
   background-color: white;
   font-size: 13px;
}

body {
  max-width: 800px;
  margin: auto;
  padding: 1em;
  line-height: 20px;
}

tt, code, pre {
   font-family: 'DejaVu Sans Mono', 'Droid Sans Mono', 'Lucida Console', Consolas, Monaco, monospace;
}

h1 {
   font-size:2.2em;
}

h2 {
   font-size:1.8em;
}

h3 {
   font-size:1.4em;
}

h4 {
   font-size:1.0em;
}

h5 {
   font-size:0.9em;
}

h6 {
   font-size:0.8em;
}

a:visited {
   color: rgb(50%, 0%, 50%);
}

pre, img {
  max-width: 100%;
}
pre {
  overflow-x: auto;
}
pre code {
   display: block; padding: 0.5em;
}

code {
  font-size: 92%;
  border: 1px solid #ccc;
}

code[class] {
  background-color: #F8F8F8;
}

table, td, th {
  border: none;
}

blockquote {
   color:#666666;
   margin:0;
   padding-left: 1em;
   border-left: 0.5em #EEE solid;
}

hr {
   height: 0px;
   border-bottom: none;
   border-top-width: thin;
   border-top-style: dotted;
   border-top-color: #999999;
}

@media print {
   * {
      background: transparent !important;
      color: black !important;
      filter:none !important;
      -ms-filter: none !important;
   }

   body {
      font-size:12pt;
      max-width:100%;
   }

   a, a:visited {
      text-decoration: underline;
   }

   hr {
      visibility: hidden;
      page-break-before: always;
   }

   pre, blockquote {
      padding-right: 1em;
      page-break-inside: avoid;
   }

   tr, img {
      page-break-inside: avoid;
   }

   img {
      max-width: 100% !important;
   }

   @page :left {
      margin: 15mm 20mm 15mm 10mm;
   }

   @page :right {
      margin: 15mm 10mm 15mm 20mm;
   }

   p, h2, h3 {
      orphans: 3; widows: 3;
   }

   h2, h3 {
      page-break-after: avoid;
   }
}
</style>



</head>

<body>
<h3>Description</h3>

<p>This scripts outlines the VCF file and the functions used to annotate and filter SNP. </p>

<hr/>

<p><strong>Code:</strong><br/>
Parent Directory:  </p>

<blockquote>
<p>&nbsp;&nbsp;&nbsp;&nbsp;/mnt/research/NMDL/2019_WB_MFM/cSNP    </p>
</blockquote>

<p>File:  </p>

<p>&nbsp;&nbsp;&nbsp;&nbsp;Function_Filter_cSNP.R  </p>

<p><strong>Input files:</strong><br/>
Directory/File  </p>

<blockquote>
<p>&nbsp;&nbsp;&nbsp;&nbsp;/mnt/research/NMDL/2019_WB_MFM/cSNP/cSNP.vcf  </p>
</blockquote>

<p><strong>Output files:</strong>  </p>

<p>Directory:  </p>

<blockquote>
<p>&nbsp;&nbsp;&nbsp;&nbsp;/mnt/research/NMDL/2019_WB_MFM/cSNP/Function_Filter_cSNP    </p>
</blockquote>

<p>Files:  </p>

<blockquote>
<p>&nbsp;&nbsp;&nbsp;&nbsp;VCF_MFM_WB.Rdata<br/>
&nbsp;&nbsp;&nbsp;&nbsp;Functions_Annotate_SNP.Rdata  </p>
</blockquote>

<p>Render R Script  </p>

<blockquote>
<p>&nbsp;&nbsp;&nbsp;&nbsp;Function_Filter_cSNP.qsub  </p>
</blockquote>

<hr/>

<h3>R Environment</h3>

<p>Load required libraries</p>

<pre><code class="r">library(vcfR)
</code></pre>

<pre><code>## 
##    *****       ***   vcfR   ***       *****
##    This is vcfR 1.8.0 
##      browseVignettes(&#39;vcfR&#39;) # Documentation
##      citation(&#39;vcfR&#39;) # Citation
##    *****       *****      *****       *****
</code></pre>

<pre><code class="r">library(foreach)
library(doParallel)
</code></pre>

<pre><code>## Loading required package: iterators
</code></pre>

<pre><code>## Loading required package: parallel
</code></pre>

<pre><code class="r">library(knitr)
</code></pre>

<p>Clear Environment</p>

<pre><code class="r">rm(list=ls())
</code></pre>

<h3>Load Data</h3>

<p>VCF cSNP File</p>

<pre><code class="r">vcf.Dir &lt;- &quot;/mnt/research/NMDL/2019_WB_MFM/cSNP&quot;
vcf &lt;- read.vcfR(paste(vcf.Dir, &quot;cSNP.vcf&quot;, sep=&quot;/&quot;), verbose = FALSE)
</code></pre>

<p>Add ID column to VCF data by concating the chromosome and position of each cSNP</p>

<pre><code class="r">vcf &lt;- addID(vcf, sep = &quot;_&quot;)
head(vcf)
</code></pre>

<pre><code>## [1] &quot;***** Object of class &#39;vcfR&#39; *****&quot;
## [1] &quot;***** Meta section *****&quot;
## [1] &quot;##fileformat=VCFv4.2&quot;
## [1] &quot;##FILTER=&lt;ID=PASS,Description=\&quot;All filters passed\&quot;&gt;&quot;
## [1] &quot;##bcftoolsVersion=1.9-64-g28bcc56+htslib-1.9-52-g6e86e38&quot;
## [1] &quot;##bcftoolsCommand=mpileup -Ou -C50 -E -Q25 -a DV,AD,ADF,ADR,SP -f /m [Truncated]&quot;
## [1] &quot;##reference=file:///mnt/research/NMDL/KER_Glycogen_and_RER_Thoroughb [Truncated]&quot;
## [1] &quot;##contig=&lt;ID=chr1,length=188260577&gt;&quot;
## [1] &quot;First 6 rows.&quot;
## [1] 
## [1] &quot;***** Fixed section *****&quot;
##      CHROM  POS   ID           REF ALT QUAL      FILTER   
## [1,] &quot;chr1&quot; &quot;229&quot; &quot;chr1_229_1&quot; &quot;C&quot; &quot;G&quot; &quot;513.0&quot;   &quot;PASS&quot;   
## [2,] &quot;chr1&quot; &quot;233&quot; &quot;chr1_233_2&quot; &quot;G&quot; &quot;T&quot; &quot;162.0&quot;   &quot;PASS&quot;   
## [3,] &quot;chr1&quot; &quot;265&quot; &quot;chr1_265_3&quot; &quot;T&quot; &quot;G&quot; &quot;763.0&quot;   &quot;PASS&quot;   
## [4,] &quot;chr1&quot; &quot;311&quot; &quot;chr1_311_4&quot; &quot;T&quot; &quot;C&quot; &quot;758.0&quot;   &quot;PASS&quot;   
## [5,] &quot;chr1&quot; &quot;363&quot; &quot;chr1_363_5&quot; &quot;G&quot; &quot;A&quot; &quot;999.0&quot;   &quot;PASS&quot;   
## [6,] &quot;chr1&quot; &quot;377&quot; &quot;chr1_377_6&quot; &quot;T&quot; &quot;C&quot; &quot;14.9975&quot; &quot;LowQual&quot;
## [1] 
## [1] &quot;***** Genotype section *****&quot;
##      FORMAT                   12065_uniq_sorted.bam         
## [1,] &quot;GT:PL:DV:SP:ADF:ADR:AD&quot; &quot;0/1:23,0,167:1:0:1,1:4,0:5,1&quot;
## [2,] &quot;GT:PL:DV:SP:ADF:ADR:AD&quot; &quot;0/0:0,18,208:0:0:2,0:4,0:6,0&quot;
## [3,] &quot;GT:PL:DV:SP:ADF:ADR:AD&quot; &quot;0/0:0,3,41:0:0:0,0:1,0:1,0&quot;  
## [4,] &quot;GT:PL:DV:SP:ADF:ADR:AD&quot; &quot;0/1:33,0,31:1:0:0,0:1,1:1,1&quot; 
## [5,] &quot;GT:PL:DV:SP:ADF:ADR:AD&quot; &quot;0/1:62,0,79:2:0:1,0:1,2:2,2&quot; 
## [6,] &quot;GT:PL:DV:SP:ADF:ADR:AD&quot; &quot;0/0:0,12,150:0:0:1,0:3,0:4,0&quot;
##      12124_uniq_sorted.bam          12130_uniq_sorted.bam         
## [1,] &quot;0/0:0,12,150:0:0:1,0:3,0:4,0&quot; &quot;1/1:165,15,0:5:0:0,3:0,2:0,5&quot;
## [2,] &quot;0/1:26,0,113:1:0:1,0:2,1:3,1&quot; &quot;0/0:0,15,170:0:0:3,0:2,0:5,0&quot;
## [3,] &quot;0/1:68,0,32:2:0:0,1:1,1:1,2&quot;  &quot;1/1:129,12,0:4:0:0,3:0,1:0,4&quot;
## [4,] &quot;0/1:24,0,110:1:0:4,1:0,0:4,1&quot; &quot;1/1:39,3,0:1:0:0,0:0,1:0,1&quot;  
## [5,] &quot;0/0:0,27,251:0:0:7,0:2,0:9,0&quot; &quot;1/1:230,24,0:8:0:0,4:0,4:0,8&quot;
## [6,] &quot;0/0:0,21,222:0:0:5,0:2,0:7,0&quot; &quot;0/0:0,27,255:0:0:5,0:4,0:9,0&quot;
##      12212_uniq_sorted.bam           12216_uniq_sorted.bam           
## [1,] &quot;0/1:145,0,192:5:0:3,2:3,3:6,5&quot; &quot;0/1:98,0,133:3:0:1,1:3,2:4,3&quot;  
## [2,] &quot;0/0:0,33,255:0:0:6,0:5,0:11,0&quot; &quot;0/0:0,21,223:0:0:2,0:5,0:7,0&quot;  
## [3,] &quot;0/1:92,0,99:3:0:1,2:2,1:3,3&quot;   &quot;0/1:54,0,110:2:4:2,0:1,2:3,2&quot;  
## [4,] &quot;0/1:81,0,183:3:0:2,1:4,2:6,3&quot;  &quot;0/1:64,0,70:2:0:1,1:1,1:2,2&quot;   
## [5,] &quot;0/1:156,0,210:6:6:3,5:4,1:7,6&quot; &quot;0/1:212,0,142:8:0:4,5:1,3:5,8&quot; 
## [6,] &quot;0/0:0,39,255:0:0:8,0:5,0:13,0&quot; &quot;0/0:0,42,255:0:0:10,0:4,0:14,0&quot;
## [1] &quot;First 6 columns only.&quot;
## [1] 
## [1] &quot;Unique GT formats:&quot;
## [1] &quot;GT:PL:DV:SP:ADF:ADR:AD&quot;
## [1]
</code></pre>

<p>Animal Ids</p>

<pre><code class="r">anim &lt;- unlist(lapply(strsplit(colnames(vcf@gt)[-1], &quot;_&quot;), function(x) x[1]))
mfm &lt;- c(&quot;9056&quot;, &quot;9066&quot;, &quot;12711&quot;, &quot;12130&quot;, &quot;12429&quot;, &quot;9054&quot;, &quot;12124&quot;, &quot;12065&quot;)
anim &lt;- data.frame(Anim=anim, MFM=ifelse(anim %in% mfm, &quot;Afected&quot;, &quot;Control&quot;))
table(anim$MFM)
</code></pre>

<pre><code>## 
## Afected Control 
##       8       8
</code></pre>

<h3>Table of VCF parameters</h3>

<h4>INFO</h4>

<pre><code class="r"># Parameter
idx &lt;- unlist(lapply(queryMETA(vcf, element=&quot;INFO&quot;), function(x) x[1]))
param &lt;- unlist(lapply(strsplit(idx, &quot;=&quot;), function(x) x[3]))

# Description
idx &lt;- unlist(lapply(queryMETA(vcf, element=&quot;INFO&quot;), function(x) x[4]))
Desc &lt;- unlist(lapply(strsplit(idx, &quot;=&quot;), function(x) x[2]))

# Table
kable(data.frame(Parameter=param, Description=Desc), caption=&quot;VCF INFO&quot;)
</code></pre>

<table><thead>
<tr>
<th align="left">Parameter</th>
<th align="left">Description</th>
</tr>
</thead><tbody>
<tr>
<td align="left">INDEL</td>
<td align="left">Indicates that the variant is an INDEL.</td>
</tr>
<tr>
<td align="left">IDV</td>
<td align="left">Maximum number of reads supporting an indel</td>
</tr>
<tr>
<td align="left">IMF</td>
<td align="left">Maximum fraction of reads supporting an indel</td>
</tr>
<tr>
<td align="left">DP</td>
<td align="left">Raw read depth</td>
</tr>
<tr>
<td align="left">VDB</td>
<td align="left">Variant Distance Bias for filtering splice-site artefacts in RNA-seq data (bigger is better)&ldquo;</td>
</tr>
<tr>
<td align="left">RPB</td>
<td align="left">Mann-Whitney U test of Read Position Bias (bigger is better)</td>
</tr>
<tr>
<td align="left">MQB</td>
<td align="left">Mann-Whitney U test of Mapping Quality Bias (bigger is better)</td>
</tr>
<tr>
<td align="left">BQB</td>
<td align="left">Mann-Whitney U test of Base Quality Bias (bigger is better)</td>
</tr>
<tr>
<td align="left">MQSB</td>
<td align="left">Mann-Whitney U test of Mapping Quality vs Strand Bias (bigger is better)</td>
</tr>
<tr>
<td align="left">SGB</td>
<td align="left">Segregation based metric.</td>
</tr>
<tr>
<td align="left">MQ0F</td>
<td align="left">Fraction of MQ0 reads (smaller is better)</td>
</tr>
<tr>
<td align="left">ICB</td>
<td align="left">Inbreeding Coefficient Binomial test (bigger is better)</td>
</tr>
<tr>
<td align="left">HOB</td>
<td align="left">Bias in the number of HOMs number (smaller is better)</td>
</tr>
<tr>
<td align="left">AC</td>
<td align="left">Allele count in genotypes for each ALT allele</td>
</tr>
<tr>
<td align="left">AN</td>
<td align="left">Total number of alleles in called genotypes</td>
</tr>
<tr>
<td align="left">DP4</td>
<td align="left">Number of high-quality ref-forward</td>
</tr>
<tr>
<td align="left">MQ</td>
<td align="left">Average mapping quality</td>
</tr>
<tr>
<td align="left">ANN</td>
<td align="left">Functional annotations: &#39;Allele &#124; Annotation &#124; Annotation_Impact &#124; Gene_Name &#124; Gene_ID &#124; Feature_Type &#124; Feature_ID &#124; Transcript_BioType &#124; Rank &#124; HGVS.c &#124; HGVS.p &#124; cDNA.pos / cDNA.length &#124; CDS.pos / CDS.length &#124; AA.pos / AA.length &#124; Distance &#124; ERRORS / WARNINGS / INFO&#39;</td>
</tr>
<tr>
<td align="left">LOF</td>
<td align="left">Predicted loss of function effects for this variant. Format: &#39;Gene_Name &#124; Gene_ID &#124; Number_of_transcripts_in_gene &#124; Percent_of_transcripts_affected&#39;</td>
</tr>
<tr>
<td align="left">NMD</td>
<td align="left">Predicted nonsense mediated decay effects for this variant. Format: &#39;Gene_Name &#124; Gene_ID &#124; Number_of_transcripts_in_gene &#124; Percent_of_transcripts_affected&#39;</td>
</tr>
</tbody></table>

<h4>FORMAT</h4>

<pre><code class="r"># Parameter  
idx &lt;- unlist(lapply(queryMETA(vcf, element=&quot;FORMAT&quot;), function(x) x[1]))
param &lt;- unlist(lapply(strsplit(idx, &quot;=&quot;), function(x) x[3]))

# Description
idx &lt;- unlist(lapply(queryMETA(vcf, element=&quot;FORMAT&quot;), function(x) x[4]))
Desc &lt;- unlist(lapply(strsplit(idx, &quot;=&quot;), function(x) x[2]))

# Table
kable(data.frame(Parameter=param, Description=Desc), caption=&quot;VCF INFO&quot;)
</code></pre>

<table><thead>
<tr>
<th align="left">Parameter</th>
<th align="left">Description</th>
</tr>
</thead><tbody>
<tr>
<td align="left">PL</td>
<td align="left">List of Phred-scaled genotype likelihoods</td>
</tr>
<tr>
<td align="left">DV</td>
<td align="left">Number of high-quality non-reference bases</td>
</tr>
<tr>
<td align="left">SP</td>
<td align="left">Phred-scaled strand bias P-value</td>
</tr>
<tr>
<td align="left">AD</td>
<td align="left">Allelic depths</td>
</tr>
<tr>
<td align="left">ADF</td>
<td align="left">Allelic depths on the forward strand</td>
</tr>
<tr>
<td align="left">ADR</td>
<td align="left">Allelic depths on the reverse strand</td>
</tr>
<tr>
<td align="left">GT</td>
<td align="left">Genotype</td>
</tr>
</tbody></table>

<h3>Filtering Function</h3>

<p>Function to merge annotation (utility function called by filterSNP)</p>

<pre><code class="r">    annotate &lt;- function(i = i, Info=Info, Ann=Ann, tmp=tmp, Des=Des, idx=idx, param=param){
        T &lt;- strsplit(tmp[[i]], &quot;;&quot;)[[1]]
        id &lt;- unlist(lapply(strsplit(T, &quot;=&quot;), function(x) x[1]))
        info &lt;- do.call(cbind, lapply(strsplit(T, &quot;=&quot;), function(x) x[2]))
        colnames(info) &lt;- id

        if(&quot;INDEL&quot; %in% id){
            info[,&quot;INDEL&quot;] &lt;- &quot;INDEL&quot;
        }

        K &lt;- data.frame(matrix(NA, nrow=1, ncol=17, dimnames=list(i, param[1:17])))
        for (j in param[1:17]){
            if (j %in% id){
                K[,j] &lt;- info[,j]
            }
        }

        if(&quot;ANN&quot; %in% id){
            ann &lt;- strsplit(info[,&quot;ANN&quot;], &quot;[|]&quot;)[[1]]
            ann &lt;- matrix(ann, nrow=length(ann)/length(idx), ncol=length(idx), byrow=TRUE, 
                dimnames=list(NULL,idx))
            ann &lt;- data.frame(SNP=rep(i, nrow(ann)), ann)
            ann$Allele &lt;- gsub(&quot;,&quot;, &quot;&quot;, ann$Allele)
        } else {
            ann &lt;- matrix(NA, nrow=1, ncol=length(idx), dimnames=list(NULL, idx))
        }

        if(&quot;LOF&quot; %in% id){
            lof &lt;- gsub(&quot;[)]&quot;, &quot;&quot;, info[,&quot;LOF&quot;])
            lof &lt;- data.frame(LOF_Numb_transcripts_in_gene=strsplit(lof, &quot;[|]&quot;)[[1]][3], 
                LOF_Percent_transcripts_affected=strsplit(lof, &quot;[|]&quot;)[[1]][4])
        } else {
            lof &lt;- data.frame(LOF_Numb_transcripts_in_gene=NA, LOF_Percent_transcripts_affected=NA)
        }

        if(&quot;NMD&quot; %in% id){
            nmd &lt;- gsub(&quot;[)]&quot;, &quot;&quot;, info[,&quot;NMD&quot;])
            nmd &lt;- data.frame(NMD_Numb_transcripts_in_gene=strsplit(nmd, &quot;[|]&quot;)[[1]][3], 
                NMD_Percent_transcripts_affected=strsplit(nmd, &quot;[|]&quot;)[[1]][4])
        } else {
            nmd &lt;- data.frame(NMD_Numb_transcripts_in_gene=NA, NMD_Percent_transcripts_affected=NA)
        }

        Info[[i]] &lt;- data.frame(data.frame(t(Des[i,])), K, lof, nmd)
        Ann[[i]] &lt;- ann

        rst &lt;- list(Info=Info[[i]], Ann=Ann[[i]])
        return(rst)
    }
</code></pre>

<p>The following function filters SNPs based on:</p>

<blockquote>
<p>Low quality: PASS filter (automatic taken from vcf)<br/>
<code>InDel</code>: removes them from downstream analysis (default=FALSE)<br/>
<code>MQSB</code>: Mann-Whitney U test of Mapping Quality vs Strand Bias p-value cutoff (default=0.05)<br/>
<code>Miss</code> : Genotype calls, filters SNP with less than <code>Miss</code> missing genotypes (default=2)<br/>
<code>TD</code> : Total allelic density: filters  SNP with less than <code>TD</code> reads (default=10)<br/>
This function can be run in parallel using <code>cl</code> to specify the number of cores (default=20)
The required input is the vcf object of class vcfR (<code>vcf</code>)and a vector of SNP to filter 
(<code>SNP</code>, optional default is NULL). When specifying SNP make sure they match the ID in the vcf.
Default for SNP is to take the IDs directly from the vcf object (Note: computationally intensive) 
 The results are formated as list with four matrices and two lists:
<code>Info</code>: SNP informtion (see INFO and FORMAT tables above for description)<br/>
<code>Ann</code> : A list containing SNP annotation (see INFO and FORMAT tables above for description)<br/>
<code>Geno</code>: Genotypes<br/>
<code>Ref</code>: Reference allele density<br/>
<code>Alt</code>: Alternative allele density<br/>
<code>FilteredInfo</code> : A list contating a summary table with the following information:<br/>
&nbsp;&nbsp;&nbsp;&nbsp;<code>FiltSum</code> : Table with the number of SNP filtered out per category<br/>
&nbsp;&nbsp;&nbsp;&nbsp;<code>FiltInfo</code> : Filteres SNP information<br/>
&nbsp;&nbsp;&nbsp;&nbsp;<code>FiltAnn</code> : Filteres SNP annotation<br/>
Filter variants Function</p>
</blockquote>

<pre><code class="r">filterSNP &lt;- function(vcf, SNP=NULL, InDel=FALSE, MQSB=0.05, Miss=2, TD=10, cl=20){
# Parameter
    idx &lt;- unlist(lapply(queryMETA(vcf, element=&quot;INFO&quot;), function(x) x[1]))
    param &lt;- unlist(lapply(strsplit(idx, &quot;=&quot;), function(x) x[3]))

# Description
    idx &lt;- unlist(lapply(queryMETA(vcf, element=&quot;INFO&quot;), function(x) x[4]))
    desc &lt;- unlist(lapply(strsplit(idx, &quot;=&quot;), function(x) x[2]))
    names(desc) &lt;- param

# SNP and animal IDs
    if(is.null(SNP)){
        SNP &lt;- vcf@fix[,&quot;ID&quot;]
    }
    ids &lt;- vcf@fix[,&quot;ID&quot;]
    anim &lt;- unlist(lapply(strsplit(colnames(vcf@gt)[-1], &quot;_&quot;), function(x) x[1]))

# SNP Information (chr, pos, etc)
    Des &lt;-  vcf@fix[,c(1,2,4,5,6,7)]
    rownames(Des) &lt;-  ids

# SNP Format
    tmp &lt;- vcf@fix[,8]
    names(tmp) &lt;- ids

# Annotation elements
    idx &lt;- strsplit(desc[[&quot;ANN&quot;]], &quot;[|]&quot;)[[1]][-16]
    idx &lt;- gsub(&quot;Functional annotations: &#39;&quot;, &quot;&quot;, idx)
    idx &lt;- gsub(&quot; &quot;, &quot;&quot;, idx)

# Merge SNP information and annotation
    Info &lt;- list()
    Ann &lt;- list()
    registerDoParallel(cl)
    rst &lt;- foreach(i=SNP) %dopar% 
        annotate(i = i, Info=Info, Ann=Ann, tmp=tmp, Des=Des, idx=idx, param=param) 

# SNP Annotation
    INFO &lt;- do.call(rbind, lapply(rst, function(x) x[[&quot;Info&quot;]]))
    ANN &lt;-  lapply(rst, function(x) x[[&quot;Ann&quot;]])
    names(ANN) &lt;- SNP

# Filter out low quality
    lowq &lt;- SNP[!INFO[,&quot;FILTER&quot;] == &quot;PASS&quot;]

# Filter out INDELs
    if (InDel == TRUE){
        indel &lt;- SNP[grep(&quot;INDEL&quot;, INFO[,&quot;INDEL&quot;])] 
    } else {
        indel &lt;- NULL
    }

# Filter out SNP with significant mapping quality vs strand bias
    N &lt;- as.numeric(INFO[,&quot;MQSB&quot;])
    names(N) &lt;- rownames(INFO)
    N[is.na(N)] &lt;- 0.04
    mqsb &lt;- names(N)[N &lt; MQSB]

# Filter based on genotype calls: less than animals with missing genotype
    gt &lt;- vcf@gt
    rownames(gt) &lt;- names(tmp)
    colnames(gt) &lt;- c(&quot;FORMAT&quot;, anim)
    gt &lt;- gt[rownames(INFO),]
    geno &lt;- apply(gt, 2, function(x) unlist(lapply(strsplit(x, &quot;:&quot;), function(y) y[1])))
    geno &lt;- geno[,-1]
    calls &lt;- rownames(geno)[apply(geno, 1, function(x) length(grep(&quot;[.]&quot;, x))) &gt; Miss]

# Filter based on base density: less than 10
    AD &lt;- apply(gt, 2, function(x) unlist(lapply(strsplit(x, &quot;:&quot;), function(y) y[7])))
    AD &lt;- AD[,-1]

# Alleleic Depth Reference
    ADR &lt;- do.call(rbind, lapply(1:nrow(AD), 
        function(x) as.numeric(unlist(lapply(strsplit(AD[x,], &quot;,&quot;), function(y) y[1])))))
    rownames(ADR) &lt;- rownames(AD)
    colnames(ADR) &lt;- colnames(AD)

# Allelic Depth Alternative
    ADA &lt;- do.call(rbind, lapply(1:nrow(AD), 
        function(x) as.numeric(unlist(lapply(strsplit(AD[x,], &quot;,&quot;), function(y) y[2])))))
    rownames(ADA) &lt;- rownames(AD)
    colnames(ADA) &lt;- colnames(AD)

# Total Allelic Density
    TAD &lt;- ADR + ADA
    tad &lt;- rownames(TAD)[apply(TAD, 1, function(x) sum(x &lt; TD)) &gt; Miss]

# Filtered SNP
    rmSNP &lt;- unique(c(lowq, indel, mqsb, calls, tad))
    SNP &lt;- SNP[!SNP %in% rmSNP]

# Table of filtered SNP
    Filt &lt;- data.frame(LowQuality=length(lowq), InDel=length(indel), MQSB=length(mqsb), 
        GenoCall=length(calls), TAD=length(tad), TotalFiltered=length(rmSNP), 
        TotalRetained=length(SNP))

# Filtered SNP information
    FiltInfo &lt;- INFO[rmSNP,]
    FiltAnn &lt;- ANN[rmSNP]
    FilteredInfo &lt;- list(FiltSum=Filt, FiltInfo=FiltInfo, FiltAnn=FiltAnn)

# Merge list of genotypes and allelic density
    rst &lt;- list(Info=INFO[SNP,], Ann=ANN[SNP], Geno=geno[SNP,], Ref=ADR[SNP,], 
        Alt=ADA[SNP,], FilteredInfo=FilteredInfo)

    return(rst)
}
</code></pre>

<p>Save VCF object to file</p>

<pre><code class="r">save(vcf, file=paste(getwd(), &quot;VCF_MFM_WB.Rdata&quot;, sep=&quot;/&quot;))
</code></pre>

<p>Save functions</p>

<pre><code class="r">save(vcf, annotate, filterSNP, file=paste(getwd(), &quot;Functions_Annotate_SNP.Rdata&quot;, sep=&quot;/&quot;))
</code></pre>

<p>Session Information</p>

<pre><code class="r">sessionInfo()
</code></pre>

<pre><code>## R version 3.5.1 (2018-07-02)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: CentOS Linux 7 (Core)
## 
## Matrix products: default
## BLAS/LAPACK: /opt/software/OpenBLAS/0.3.1-GCC-7.3.0-2.30/lib/libopenblas_sandybridgep-r0.3.1.so
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] parallel  stats     graphics  grDevices utils     datasets  methods  
## [8] base     
## 
## other attached packages:
## [1] doParallel_1.0.11 iterators_1.0.10  foreach_1.4.4     vcfR_1.8.0       
## [5] knitr_1.20       
## 
## loaded via a namespace (and not attached):
##  [1] Rcpp_1.0.1        cluster_2.0.7-1   magrittr_1.5     
##  [4] MASS_7.3-51.1     viridisLite_0.3.0 ape_5.1          
##  [7] lattice_0.20-38   highr_0.7         stringr_1.4.0    
## [10] tools_3.5.1       grid_3.5.1        nlme_3.1-137     
## [13] mgcv_1.8-24       vegan_2.5-4       permute_0.9-5    
## [16] Matrix_1.2-14     codetools_0.2-15  evaluate_0.12    
## [19] pinfsc50_1.1.0    stringi_1.4.3     compiler_3.5.1   
## [22] memuse_4.0-0
</code></pre>

<h3>Run R Script</h3>

<pre><code class="r"># htmlRunR
# Function_Filter_cSNP.R nodes=1,cpus-per-task=1,time=01:00:00,mem=10G +Function to filter and annotate SNP
</code></pre>

</body>

</html>
